Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+{
+    /// <summary>
+    /// A distributed cache that stores entries in Firestore.
+    /// </summary>
+    public class FirestoreCache : IDistributedCache
+    {
+        private readonly FirestoreDb _firestore;
+        private readonly CollectionReference _cacheEntries;
+        private readonly ILogger<FirestoreCache> _logger;
+
+        public FirestoreCache(FirestoreDb firestore, ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+        {
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if (string.IsNullOrWhiteSpace(collection))
+            {
+                throw new ArgumentException("Must not be empty",
+                    nameof(collection));
+            }
+
+            _firestore = firestore;
+            _cacheEntries = _firestore.Collection(collection);
+            _logger = logger;
+        }
+        public FirestoreCache(string projectId, ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+            : this(FirestoreDb.Create(projectId), logger, collection)
+        {
+        }
+
+        public FirestoreCache(ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+            : this(FirestoreDb.Create(GetProjectId()), logger, collection)
+        {            
+        }
+
+        byte[] IDistributedCache.Get(string key) =>
+            ValueFromSnapshot(_cacheEntries.Document(key).GetSnapshotAsync().Result);
+
+        async Task<byte[]> IDistributedCache.GetAsync(string key, CancellationToken token) =>
+            ValueFromSnapshot(await _cacheEntries.Document(key).GetSnapshotAsync(token));
+
+        /// <summary>
+        /// Extracts the cache value from a firestore document snapshot.
+        /// Checks all the expiration fields so never returns a stale value.
+        /// </summary>
+        /// <param name="snapshot">The snapshot to pull the cache value from.</param>
+        /// <returns>The cache value or null.</returns>
+        byte[] ValueFromSnapshot(DocumentSnapshot snapshot)
+        {
+            if (!snapshot.Exists)
+            {
+                return null;
+            }
+            CacheDoc doc = snapshot.ConvertTo<CacheDoc>();
+            var now = DateTime.UtcNow;
+            if (doc.AbsoluteExpiration.HasValue &&
+                doc.AbsoluteExpiration.Value < now)
+            {
+                return null;
+            }
+            var slidingExpiration = doc.LastRefresh.GetValueOrDefault()
+                + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.GetValueOrDefault());
+            if (slidingExpiration < now)
+            {
+                return null;
+            }
+            return doc.Value;
+        }
+        void IDistributedCache.Refresh(string key)
+        {
+            try
+            {
+                _cacheEntries.Document(key)

```

Comment:
Possibly unwrap to a single line?

Reply:
The line would be > 80 chars.  What is the limit?